bpfman/src/lib.rs

```diff
@@ -1249,35 +1279,184 @@ pub(crate) fn add_single_attach_program(root_db: &Db, p: &mut Program) -> Result
 
             let id = program.data.get_id()?;
 
-            let program_pin_path = format!("{RTDIR_FS}/prog_{}", id);
-            let fn_name = program.get_fn_name()?;
+            let program_pin_path = format!("{RTDIR_FS}/prog_{id}");
 
             uprobe
                 .pin(program_pin_path.clone())
                 .map_err(BpfmanError::UnableToPinProgram)?;
 
-            match program.get_container_pid()? {
+            Ok(id)
+        }
+        Program::Fentry(ref mut program) => {
+            let fn_name = program.get_fn_name()?;
+            let btf = Btf::from_sys_fs()?;
+            let fentry: &mut FEntry = raw_program.try_into()?;
+            fentry
+                .load(&fn_name, &btf)
+                .map_err(BpfmanError::BpfProgramError)?;
+            program.get_data_mut().set_kernel_info(&fentry.info()?)?;
+
+            let id = program.data.get_id()?;
+
+            fentry
+                .pin(format!("{RTDIR_FS}/prog_{id}"))
+                .map_err(BpfmanError::UnableToPinProgram)?;
+
+            Ok(id)
+        }
+        Program::Fexit(ref mut program) => {
+            let fn_name = program.get_fn_name()?;
+            let btf = Btf::from_sys_fs()?;
+            let fexit: &mut FExit = raw_program.try_into()?;
+            fexit
+                .load(&fn_name, &btf)
+                .map_err(BpfmanError::BpfProgramError)?;
+            program.get_data_mut().set_kernel_info(&fexit.info()?)?;
+
+            let id = program.data.get_id()?;
+
+            fexit
+                .pin(format!("{RTDIR_FS}/prog_{id}"))
+                .map_err(BpfmanError::UnableToPinProgram)?;
+
+            Ok(id)
+        }
+        Program::Tcx(ref mut program) => {
+            debug!("Loading TCX program");
+            let tcx: &mut SchedClassifier = raw_program.try_into()?;
+
+            debug!("Calling load on TCX program");
+            tcx.load()?;
+            program.get_data_mut().set_kernel_info(&tcx.info()?)?;
+
+            let id = program.data.get_id()?;
+
+            debug!("Pinning TCX program");
+            tcx.pin(format!("{RTDIR_FS}/prog_{id}"))
```

Comment:
Should we use a function for the program and link pin paths instead of raw format! calls?

Reply:
#1430